primitives/core/src/lib.rs

```diff
@@ -118,12 +118,23 @@ pub mod well_known_keys {
 }
 
 /// Something that should be called when a downward message is received.
-pub trait DownwardMessageHandler {
-	/// Handle the given downward message.
-	fn handle_downward_message(msg: InboundDownwardMessage) -> Weight;
+pub trait DmpMessageHandler {
+	/// Handle some incoming DMP messages (note these are individual XCM messages).
+	///
+	/// Also, process messages up to some `max_weight`.
+	fn handle_dmp_messages(
+		iter: impl Iterator<Item=(RelayBlockNumber, Vec<u8>)>,
+		max_weight: Weight,
+	) -> Weight;
 }
-impl DownwardMessageHandler for () {
-	fn handle_downward_message(_msg: InboundDownwardMessage) -> Weight { 0 }
+impl DmpMessageHandler for () {
+	fn handle_dmp_messages(
+		iter: impl Iterator<Item=(RelayBlockNumber, Vec<u8>)>,
+		_max_weight: Weight,
+	) -> Weight {
+		for _ in iter {}
```

Comment:
maybe dumb question, but this gets optimized away?

Reply:
I don't think so as `next()` can have side effects.